File: autonav_ws/src/autonav_libs/src/node.cpp
#include "autonav_libs/node.h"
#include <unistd.h>

#define UNUSED(expr) do { (void)(expr); } while (0)

namespace Autonav
{
	AutoNode::AutoNode(std::string node_name) : Node(node_name)
	{
		id = Autonav::hash(node_name);

		// State System
		deviceStates[getDeviceID()] = DeviceState::OFF;
		systemStateSubscriber = this->create_subscription<autonav_msgs::msg::SystemState>("/autonav/state/system", 10, std::bind(&AutoNode::onSystemState, this, std::placeholders::_1));
		deviceStateSubscriber = this->create_subscription<autonav_msgs::msg::DeviceState>("/autonav/state/device", 10, std::bind(&AutoNode::onDeviceState, this, std::placeholders::_1));
		deviceStateClient = this->create_client<autonav_msgs::srv::SetDeviceState>("/autonav/state/set_device_state");
		systemStateClient = this->create_client<autonav_msgs::srv::SetSystemState>("/autonav/state/set_system_state");

		// Configuration
		auto configurationSubscriber = this->create_subscription<autonav_msgs::msg::ConfigurationInstruction>("/autonav/configuration", 10, std::bind(&Configuration::onConfigurationInstruction, &config, std::placeholders::_1));
		auto configurationPublisher = this->create_publisher<autonav_msgs::msg::ConfigurationInstruction>("/autonav/configuration", 10);
		config = Configuration(id, configurationSubscriber, configurationPublisher);
	}

	AutoNode::~AutoNode()
	{
	}

	int64_t AutoNode::getDeviceID()
	{
		return id;
	}

	void AutoNode::setEStop(bool state)
	{
		auto request = std::make_shared<autonav_msgs::srv::SetSystemState::Request>();
		request->state = this->state.state;
		request->estop = state;
		request->mobility = this->state.mobility;
		systemStateClient->async_send_request(request);
	}

	void AutoNode::setMobility(bool state)
	{
		auto request = std::make_shared<autonav_msgs::srv::SetSystemState::Request>();
		request->state = this->state.state;
		request->estop = this->state.estop;
		request->mobility = state;
		systemStateClient->async_send_request(request);
	}

	void AutoNode::setSystemState(SystemState state)
	{
		// Send a system state client request
		auto request = std::make_shared<autonav_msgs::srv::SetSystemState::Request>();
		request->state = static_cast<uint8_t>(state);
		request->estop = this->state.estop;
		request->mobility = this->state.mobility;
		systemStateClient->async_send_request(request);
	}

	void AutoNode::setDeviceState(DeviceState state)
	{
		auto request = std::make_shared<autonav_msgs::srv::SetDeviceState::Request>();
		request->device = getDeviceID();
		request->state = static_cast<uint8_t>(state);
		deviceStateClient->async_send_request(request);
	}

	void AutoNode::onSystemState(const autonav_msgs::msg::SystemState::SharedPtr msg)
	{
		auto old = state;
		state = *msg;
		transition(old, state);
	}

	void AutoNode::onDeviceState(const autonav_msgs::msg::DeviceState::SharedPtr msg)
	{
		deviceStates[msg->device] = static_cast<DeviceState>(msg->state);
		if (msg->device != getDeviceID())
		{
			return;
		}

		if (msg->state == static_cast<uint8_t>(DeviceState::STANDBY))
		{
			config.recache();
			configure();
			return;
		}
	}

	void AutoNode::transition(autonav_msgs::msg::SystemState old, autonav_msgs::msg::SystemState updated)
	{
		UNUSED(old);
		switch(updated.state)
		{
			case SystemState::SHUTDOWN:
				break;
			case SystemState::DISABLED:
				if (getDeviceState() == DeviceState::OPERATING)
				{
					setDeviceState(DeviceState::READY);
				}
				break;
			case SystemState::MANUAL:
			case SystemState::AUTONOMOUS:
				if (getDeviceState() == DeviceState::READY)
				{
					setDeviceState(DeviceState::OPERATING);
				}
				break;
		}
	}

	autonav_msgs::msg::SystemState AutoNode::getSystemState()
	{
		return state;
	}

	DeviceState AutoNode::getDeviceState()
	{
		return deviceStates[getDeviceID()];
	}

	DeviceState AutoNode::getDeviceState(std::string id)
	{
		return deviceStates[Autonav::hash(id)];
	}

	std::map<int64_t, DeviceState> AutoNode::getDeviceStates()
	{
		return deviceStates;
	}
}

File: autonav_ws/src/autonav_libs/autonav_libs/node.py
from autonav_libs.state import DeviceStateEnum, SystemStateEnum
from autonav_msgs.srv import SetDeviceState, SetSystemState
from autonav_msgs.msg import DeviceState, SystemState
from autonav_libs.configuration import Configuration
from autonav_libs import hash
from rclpy.node import Node
import os
import signal


class AutoNode(Node):
    def __init__(self, node_name):
        super().__init__(node_name)
        self.id = hash(node_name)

        # State System
        self.config = Configuration(node_name, self)
        self.deviceStateSubscriber = self.create_subscription(
            DeviceState, "/autonav/state/device", self.onDeviceState, 10)
        self.systemStateSubscriber = self.create_subscription(
            SystemState, "/autonav/state/system", self.onSystemState, 10)
        self.deviceStateClient = self.create_client(
            SetDeviceState, "/autonav/state/set_device_state")
        self.systemStateClient = self.create_client(
            SetSystemState, "/autonav/state/set_system_state")

        # Configuration
        self.config = Configuration(self.id, self)
        self.deviceStates = {}
        self.state = SystemState()

    def configure(self):
        pass

    def getDeviceID(self) -> int:
        return self.id

    def setSystemStateInternal(self, state: SystemState):
        request = SetSystemState.Request()
        request.state = state.state
        request.estop = state.estop
        request.mobility = state.mobility
        self.systemStateClient.call_async(request)

    def setDeviceState(self, state: DeviceStateEnum):
        request = SetDeviceState.Request()
        request.state = state
        request.device = self.id
        self.deviceStateClient.call_async(request)

    def setSystemState(self, state: SystemStateEnum):
        self.setSystemStateInternal(SystemState(
            state = state,
            estop = self.state.estop,
            mobility = self.state.mobility
        ))
    
    def setEStop(self, state: bool):
        self.setSystemStateInternal(SystemState(
            state = self.state.state,
            estop = state,
            mobility = self.state.mobility
        ))

    def setMobility(self, state: bool):
        self.setSystemStateInternal(SystemState(
            state = self.state.state,
            estop = self.state.estop,
            mobility = state
        ))

    def onSystemState(self, state: SystemState):
        if state.state == SystemStateEnum.SHUTDOWN:
            os.kill(os.getpid(), signal.SIGINT)

        old = self.state
        self.state = state
        self.transition(old, state)

    def onDeviceState(self, state: DeviceState):
        self.deviceStates[state.device] = state.state
        if state.device != self.id:
            return
        
        if state.state == DeviceStateEnum.STANDBY:
            self.config.recache()
            self.configure()
            return

    def transition(self, _: SystemState, updated: SystemState):
        if updated.state == SystemStateEnum.DISABLED:
            return
        
        if updated.state == SystemStateEnum.DISABLED and self.getDeviceState() == DeviceStateEnum.OPERATING:
            self.setDeviceState(DeviceStateEnum.READY)
            return
        
        if (updated.state == SystemStateEnum.MANUAL or updated.state == SystemStateEnum.AUTONOMOUS) and self.getDeviceState() == DeviceStateEnum.READY:
            self.setDeviceState(DeviceStateEnum.OPERATING)
            return

    def getSystemState(self) -> SystemState:
        return self.state

    def getDeviceState(self, id: int = None) -> DeviceStateEnum:
        return self.deviceStates[id] if id else self.deviceStates[self.id]
    
    def getDeviceStates(self):
        return self.deviceStates
